### PR TITLE
Uxd 374 list box tags

### PR DESCRIPTION
**Pull request by nahumzs**
#### Uncontrolled example:

```jsx
import React from "react";
import ListBoxWithTags, { useListBoxWithTags } from "@paprika/list-box-with-tags";
import animals from "../mocks";

const defaultFilteredData = animals.slice(0, 20);
const defaultData = animals;

export default function App() {

  // All props are provided by useListBoxWithTags
  const { isSelected, filteredData, getSelectedOptions, ...moreUseListBoxWithTagsProps } = useListBoxWithTags("label", {
    defaultData,
    defaultFilteredData,
  });

  return (
    <div style={{ padding: "32px" }}>
      <ListBoxWithTags selectedOptions={getSelectedOptions()} {...moreUseListBoxWithTagsProps}>
        {filteredData.map(option => {
          if (typeof option.isCustom !== "undefined") {
            return null;
          }

          return !isSelected(option.label) ? (
            <ListBoxWithTags.Option value={option.label} key={option.label} label={option.label}>
              {option.label}
            </ListBoxWithTags.Option>
          ) : null;
        })}
      </ListBoxWithTags>
    </div>
  );
}

// Note
// At any point you can execute `getSelectedOptions()` which returns an array of the current selected options.
```

Will produce:

#### useListBoxWithTags hook

To make the component uncontrolled the ListBoxWithTags provides a hook

`useListBoxWithTags` receive two parameters the first one is the `key` property from your defaultData items, which let the hook to know what's the unique key to use for storing selected items and other internal features.

The second parameter received and object with the following properties: 

```
defaultData = []

it is used to store you default Data and is an object array, it's expected that at least those object included a "label:string" property.
```

```
defaultFilteredData = []

A subset of `defaultData` which is used to store the current filtered items once the user type on the filter.
```

```
defaultSelectedKeys = [] 

An array with keys that can be used to prefill the component with selected items.
```

```
filterAttribute = "label"

in case you want to make use of another property to filter your results. 
```

It will return the props you need to fill your ListBoxWithTags, ex:

```js
  const { isSelected, filteredData, getSelectedOptions, ...moreUseListBoxWithTagsProps } = useListBoxWithTags("label", {
    defaultData,
    defaultFilteredData,
  });
```

#### Custom Pill

You can take full control when rendering the Pills to adapt it to your needs:

```jsx
  function renderPill({ option, Pill, onRemove }) {
    const color = getAvatarColors(option.label);
    return (
      <Pill onRemove={onRemove} key={option.label}>
        <div style={{ display: "flex", justifyContent: "center", alignItems: "center" }}>
          <div style={styleForPill(color)}>{option.label.substring(0, 2)}</div>
          {option.label}
        </div>
      </Pill>
    );
  }
```

#### Prop Table 

| Prop              | Type       | required | default        | Description                                                                                                       |
| ----------------- | ---------- | -------- | -------------- | ----------------------------------------------------------------------------------------------------------------- |
| children          | node       | true     | -              | Expect <ListBoxWithTags.Option />                                                                                 |
| filter            | func       | false    | undefined      | filter function for the ListBoxWithTags can be pair with ListBoxWithTags.filter                                   |
| noResultsMessage  | node       | false    | null           | String message to be display when there are not results                                                           |
| onChange          | func       | false    | () => {}       | Callback whenever the user change a selection on the ListBoxWithTags                                              |
| onCustomOption    | func       | false    | null           | Callback whenever the user input a new custom option like some@email.com, pass undefined to ignore this behaviour |
| onRemove          | func       | false    | () => {}       | Callback once a pill is remove from the Trigger                                                                   |
| regexCustomOption | instanceOf | false    | /^.+@.+\..+\$/ | Regex that match the input of the user and reports to onCustomOption. The default is a basic email regex          |
| renderPill        | func       | false    | null           | Render prop to override the default Pill style, see example for it's uses.                                        |
| selectedOptions   | arrayOf    | false    | null           | An array of id that helps the ListBoxWithTags to known what elements are selected                                 |

#### Controlled component
This is a more contrived example but shows you how to do a Fully controlled component.

```jsx
import React from "react";
import ListBoxWithTags from "../../src";
import animals from "../mocks";

// please use a proper debounce function in a your application
const debounce = (func, wait) => {
  let idTimeout = null;
  return (...args) => {
    clearTimeout(idTimeout);
    idTimeout = setTimeout(() => {
      func(...args);
    }, wait);
  };
};

const animalSubset = animals.slice(0, 20);

export default function Controlled() {
  const [data] = React.useState(animalSubset);

  const [selectedKeys, setSelectedKeys] = React.useState(["Alpaca"]);
  const [dataFiltered, setDataFiltered] = React.useState(data);

  function checkIfIsSelected(id) {
    return selectedKeys.includes(id);
  }

  function handleChange(option, options, selectedOption) {
    setSelectedKeys(prev => {
      const prevClone = prev.slice(0);
      const id = options[selectedOption].value;
      if (prev.includes(id)) {
        prev.splice(prevClone.indexOf(id), 1);
        return prev;
      }

      return [...new Set(prevClone.concat([id]))];
    });

    setDataFiltered(data);
  }

  function handleRemove(option) {
    const index = selectedKeys.indexOf(option.label);
    if (index >= 0) {
      setSelectedKeys(prev => {
        const clone = prev.slice(0);
        clone.splice(index, 1);
        return clone;
      });
    }
  }

  function handleAddedOption(label) {
    const option = { label, isCustom: true };
    // this is just an example
    // but in real life you could have a list/dictionary with the
    // value of the selected items where you could push the new value
    animals.push(option);

    setDataFiltered(() => data);

    setSelectedKeys(prev => {
      return prev.concat(option.label);
    });
  }

  const handleFilterDebounce = React.useMemo(
    () =>
      debounce(({ search }) => {
        if (search === "") {
          setDataFiltered(data);
          return;
        }

        const result = ListBoxWithTags.filter(search, animals);
        setDataFiltered(result);
      }, 250),
    [data]
  );

  return (
    <div style={{ padding: "32px" }}>
      <ListBoxWithTags
        noResultsMessage="No results found, but you can add an email and then press enter..."
        onChange={handleChange}
        onCustomOption={handleAddedOption}
        onRemove={handleRemove}
        selectedOptions={selectedKeys.length ? animals.filter(item => selectedKeys.includes(item.label)) : []}
        filter={handleFilterDebounce}
      >
        {dataFiltered.map(option => {
          if (typeof option.isCustom !== "undefined") {
            return null;
          }

          return !checkIfIsSelected(option.label) ? (
            <ListBoxWithTags.Option value={option.label} key={option.label} label={option.label}>
              {option.label}
            </ListBoxWithTags.Option>
          ) : null;
        })}
      </ListBoxWithTags>
    </div>
  );
}

```

### Updates 📦
`new`

### Storybook 📕
http://storybooks.highbond-s3.com/paprika/UXD-374-ListBoxTags

### Screenshots 📸
_optional but highly recommended_

### References 🔗
_relevant Jira ticket / GitHub issues_

### CODE REVIEW

**Comment by nahumzs** [comment]
Note: Written documentation and split for Pill in their own component in the following PR, Don't want this one to grow too much.

File: packages/ListBoxWithTags/package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@paprika/list-box-with-tags",
+  "version": "0.0.1",
+  "description": "ListBoxWithTags behave as a ListBox with the difference that the selected options stacked on pills next to each other",
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/package.json:4]: Since there is no `Tag` component, maybe this should be `ListBoxWithPills`?
Maybe they dont know about the `ListBox`, so we shouldn't reference it in the description?

```
The ListBoxWithTags component allows users to select one or more options from a list contained within the component, showing a pill next to each.
```

As discussed, maybe we don't even need a separate component... it's still just a `ListBox`

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import PropTypes from "prop-types";
+import ListBox from "@paprika/list-box";
+import useI18n from "@paprika/l10n/lib/useI18n";
+import Pill, { Pills } from "./components/Pill";
+import { filter } from "./helpers";
+import * as sc from "./ListBoxWithTags.styles";
+
+const propTypes = {
+  /** Expect <ListBoxWithTags.Option /> */
+  children: PropTypes.node.isRequired,
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:11]: `PropTypes.instanceOf(ListBoxWithTags.Option)`

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  /** filter function for the ListBoxWithTags can be pair with ListBoxWithTags.filter  */
+  filter: PropTypes.func,
+  /** String message to be display when there are not results  */
+  noResultsMessage: PropTypes.node,
+  /** Callback whenever the user change a selection on the ListBoxWithTags  */
+  onChange: PropTypes.func,
+  /** Callback whenever the user input a new custom option like some@email.com, pass undefined to ignore this behaviour */
+  onCustomOption: PropTypes.func,
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:19]: I think `onAddCustomOption` is more clear

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  /** Callback once a pill is remove from the Trigger */
+  onRemove: PropTypes.func,
+  /** Regex that match the input of the user and reports to onCustomOption. The default is a basic email regex */
+  regexCustomOption: PropTypes.instanceOf(RegExp),
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:23]: I think it should be `customOptionRegex` or `customOptionRegexPattern`

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  /** Render prop to override the default Pill style, see example for it's uses.  */
+  renderPill: PropTypes.func,
+  /** An array of id that helps the ListBoxWithTags to known what elements are selected  */
+  selectedOptions: PropTypes.arrayOf(PropTypes.shape({})),
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:27]: `arrayOf(PropTypes.object)`?

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+};
+
+const defaultProps = {
+  filter: undefined,
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:31]: Why `undefined` and not `null`?

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    renderPill,
+    selectedOptions,
+    ...moreProps
+  } = props;
+  const { t } = useI18n();
+
+  const refDivRoot = React.useRef(null);
+  /* eslint-disable react/prop-types */
+  const size =
+    typeof props.size !== "undefined" &&
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:109]: Why not just put it in `defaultProps`?

File: packages/ListBoxWithTags/src/ListBoxWithTags.js [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    [ListBox.types.size.MEDIUM, ListBox.types.size.SMALL, ListBox.types.size.LARGE].includes(props.size)
+      ? props.size
+      : ListBox.types.size.MEDIUM;
+  /* eslint-enable react/prop-types */
+
+  const refFilter = React.useRef(null);
+
+  function handleKeyDown(event) {
+    const label = event.target.value;
+    const regexEmail = regexCustomOption;
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.js:119]: dont need `regexEmail` variable

File: packages/ListBoxWithTags/src/ListBoxWithTags.styles.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  `;
+});
+
+export const PlaceHolderText = styled.div(() => {
+  return css`
+    border-radius: ${tokens.border.radius};
+    line-height: 1;
+    margin-bottom: ${tokens.spaceSm};
+    margin-right: ${tokens.spaceSm};
+    padding: ${tokens.spaceSm[0] / 2}px ${tokens.spaceSm};
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/ListBoxWithTags.styles.js:54]: why is `spaceSm` an array...?

File: packages/ListBoxWithTags/src/components/Pill/Pill.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import PropTypes from "prop-types";
+import Close from "@paprika/icon/lib/Times";
+import * as sc from "./Pill.styles";
+
+const propTypes = {
+  onRemove: PropTypes.func.isRequired,
+  children: PropTypes.node.isRequired,
+  size: PropTypes.oneOf(["small", "medium", "large"]),
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/components/Pill/Pill.js:9]: Constants. On lines 13 and 39, too.

File: packages/ListBoxWithTags/src/helpers.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+  const regex = new RegExp(`${search}`, "gi");
+  const results = [];
+  for (let i = 0, len = data.length; i < len; i++) {
+    if (data[i][id].match(regex) !== null) {
+      results.push(data[i]);
+    }
+  }
+
+  return results;
+};
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/helpers.js:11]: I think this is just:

```
export const filter = (pattern, data, id = "label") => {
  const regex = new RegExp(pattern, "gi");  
  return data.filter(datum => datum[id].match(regex) !== null)
};
```

File: packages/ListBoxWithTags/src/useListBoxWithTags.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+      return [...new Set(prevClone.concat([id]))];
+    });
+
+    setFilteredData(defaultData);
+  }
+
+  function getSelectedOptions() {
+    if (selectedKeys.length) {
+      const options = [];
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/useListBoxWithTags.js:48]: Put `options` outside of the `if` statement, and at the end of this function `return options;`

File: packages/ListBoxWithTags/src/useListBoxWithTags.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        return clone;
+      });
+    }
+  }
+
+  function isSelected(id) {
+    return selectedKeys.includes(id);
+  }
+
+  const handleCustomOption = (func = label => ({ label, isCustom: true })) => label => {
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/useListBoxWithTags.js:74]: 😨

File: packages/ListBoxWithTags/stories/examples/UncontrolledUser.js [hunk 14]
```diff
@@ -0,0 +1,166 @@
+import React from "react";
+import Popover from "@paprika/popover";
+import { getAvatarColors } from "@paprika/avatar/lib/helpers";
+import Avatar from "@paprika/avatar";
+import ListBoxWithTags, { useListBoxWithTags } from "../../src";
+
+// prettier-ignore
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/stories/examples/UncontrolledUser.js:7]: This should be extracted to a separate file, and prettified

File: packages/ListBoxWithTags/src/useListBoxWithTags.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+  data.forEach(d => {
+    dictionary[d[key]] = d;
+  });
+
+  return dictionary;
+}
+
+export default function useListBoxWithTags(
+  key,
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/src/useListBoxWithTags.js:16]: What do you think about putting `key` in the object, too?

File: packages/ListBoxWithTags/test/ListBoxWithTags.spec.js [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+      { id: 2, label: "The Fox" },
+      { id: 3, label: "The Raccoon" },
+      { id: 4, label: "Dinosaur" },
+    ];
+
+    expect(ListBoxTags.filter("dog", data)).toMatchSnapshot();
+    expect(ListBoxTags.filter("fox", data)).toMatchSnapshot();
+    expect(ListBoxTags.filter("Raccoon", data)).toMatchSnapshot();
+    expect(ListBoxTags.filter("the", data)).toMatchSnapshot();
+    expect(ListBoxTags.filter("saur", data)).toMatchSnapshot();
```
Comment (jamiek-acl) [inline packages/ListBoxWithTags/test/ListBoxWithTags.spec.js:45]: I thought we weren't using snapshots anymore

File: packages/ListBox/src/hooks/hooks.js [hunk 17]
```diff
@@ -111,43 +111,40 @@ export function useOnScrolled() {
...
       const rectOption = $option.getBoundingClientRect();
 
-      // if options its between the box don't do anything
-      if (rectOption.top >= rectBox.top && rectOption.bottom <= rectBox.bottom) {
+      const nextTopPositionForOption =
+        rectOption.height * state.activeOption + state.activeOption * (marginBottomNumber + marginTopNumber);
+
+      // this occurs when the first element is a Divider so we want to scroll to the top
+      // so the user can see the Divider text instead to see only the option
+      if (state.activeOption - 1 === 0 && state.options[state.activeOption - 1].isDisabled) {
```
[the remark this answers is not in the document]
Comment (nahumzs) [reply packages/ListBox/src/hooks/hooks.js:125]: 🤔 it might be that all of them are disabled but at that point, you couldn't use the up/down arrow, is common to have the first element to be a Divider (group) and unlikely you have two separators at the beginning. Could happen just doesn't make sense from UI/UX perspective.

File: packages/ListBoxWithTags/stories/mocks.js [hunk 18]
```diff
@@ -0,0 +1,2 @@
+// prettier-ignore
```
[the remark this answers is not in the document]
Comment (nahumzs) [reply packages/ListBoxWithTags/stories/mocks.js]: so it doesn't count as lines and make my PR looks bigger haha when is not :P

File: packages/ListBoxWithTags/package.json (hunk 1 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/package.json]: Remove ListBowWithTags as a package and put it inside of ListBox, the dev experience has increased I think.

File: packages/ListBoxWithTags/src/ListBoxWithTags.js (hunk 5 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/ListBoxWithTags.js]: yeap the user can have any kind of object data on it. Can't validate that :P

File: packages/ListBoxWithTags/src/ListBoxWithTags.js (hunk 6 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/ListBoxWithTags.js]: if you set null the filter prop will be spread if you put undefined the filter prop will be removed by react and will not be spread.

File: packages/ListBoxWithTags/src/ListBoxWithTags.js (hunk 7 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/ListBoxWithTags.js]: Ohh this is so I don't expose this prop because is already part of the ListBox.

File: packages/ListBoxWithTags/stories/mocks.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (jamiek-acl) [reply packages/ListBoxWithTags/stories/mocks.js]: I think that ship has sailed

File: packages/ListBox/src/components/WithTags/WithTags.js [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+  /** Callback whenever the user input a new custom option like some@email.com, pass undefined to ignore this behaviour */
+  onCustomOption: PropTypes.func,
+  /** Callback once a pill is remove from the Trigger */
+  onRemove: PropTypes.func,
+  /** Regex that match the input of the user and reports to onCustomOption. The default is a basic email regex */
+  regexCustomOption: PropTypes.instanceOf(RegExp),
+  /** Render prop to override the default Pill style, see example for it's uses.  */
+  renderPill: PropTypes.func,
+  /** An array of id that helps the ListBoxWithTags to known what elements are selected  */
+  selectedOptions: PropTypes.arrayOf(PropTypes.shape({})),
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/WithTags.js:28]:
```suggestion
  selectedOptions: PropTypes.arrayOf(PropTypes.object),
```

File: packages/ListBox/src/components/WithTags/WithTags.js [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    selectedOptions,
+    ...moreProps
+  } = props;
+  const { t } = useI18n();
+
+  const refDivRoot = React.useRef(null);
+  /* eslint-disable react/prop-types */
+  const size =
+    typeof props.size !== "undefined" &&
+    [ListBox.types.size.MEDIUM, ListBox.types.size.SMALL, ListBox.types.size.LARGE].includes(props.size)
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/WithTags.js:111]:
```suggestion
    [Object.keys(ListBox.types.size)].includes(props.size)
```

File: packages/ListBox/src/components/WithTags/WithTags.js [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+  selectedOptions: null,
+};
+
+const renderTrigger = ({ t, size, selectedOptions, onRemove, renderPill }) => (...args) => {
+  const [, , , attributes] = args;
+  const { propsForTrigger, refTrigger, dispatch, types, handleKeyDown, handleKeyUp } = attributes;
+
+  function handleClick(event) {
+    event.stopPropagation();
+    if (event.target.dataset.pkaAnchor === "listbox-tags-pill-delete") {
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/WithTags.js:48]:
```suggestion
    if (event.target.dataset.pkaAnchor === "list-box-tags.pill.delete") {
```

File: packages/ListBox/src/components/WithTags/WithTags.js [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+    [ListBox.types.size.MEDIUM, ListBox.types.size.SMALL, ListBox.types.size.LARGE].includes(props.size)
+      ? props.size
+      : ListBox.types.size.MEDIUM;
+  /* eslint-enable react/prop-types */
+
+  const refFilter = React.useRef(null);
+
+  function handleKeyDown(event) {
+    const label = event.target.value;
+    const regexEmail = regexCustomOption;
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/WithTags.js:120]: You don't need the variable `regexEmail`, and who says it will always be an `email` regular expression?

File: packages/ListBox/src/helpers/filter.js [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+  const regex = new RegExp(`${search}`, "gi");
+  const results = [];
+  for (let i = 0, len = data.length; i < len; i++) {
+    if (data[i][id].match(regex) !== null) {
+      results.push(data[i]);
+    }
+  }
+
+  return results;
+};
```
Comment (jamiek-acl) [inline packages/ListBox/src/helpers/filter.js:11]:
```suggestion
export const filter = (pattern, data, id = "label") => {
  const regex = new RegExp(pattern, "gi");  
  return data.filter(datum => datum[id].match(regex) !== null)
};
```

File: packages/ListBox/src/hooks/hooks.js (hunk 17 above)
Comment (jamiek-acl) [inline packages/ListBox/src/hooks/hooks.js:125]:
```suggestion
      if (state.activeOption === 1 && state.options[0].isDisabled) {
```

File: packages/ListBox/src/components/WithTags/useWithTags.js [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+      const options = [];
+      selectedKeys.forEach(key => {
+        options.push(dataDictionary[key]);
+      });
+
+      return options;
+    }
+
+    return [];
+  }
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/useWithTags.js:56]:
```suggestion
  function getSelectedOptions() {
    return selectedKeys.map(key => dataDictionary[key]);
  }
```

File: packages/ListBox/src/components/WithTags/useWithTags.js [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+    const index = selectedKeys.indexOf(option[key]);
+    if (index >= 0) {
+      setSelectedKeys(prev => {
+        const clone = prev.slice(0);
+        clone.splice(index, 1);
+        return clone;
+      });
+    }
+  }
+
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/WithTags/useWithTags.js:68]: Can `.filter` be used here?

File: packages/ListBox/src/components/WithTags/WithTags.js [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+      {...propsForTrigger()}
+      onClick={handleClick}
+      onKeyUp={handleKeyUp}
+      onKeyDown={handleKeyDown}
+      size={size}
+    >
+      <Pills>
+        {selectedOptions.map(item => {
+          if (typeof renderPill === "function") {
+            return renderPill({ option: item, Pill, onRemove: handleRemove(item) });
```
Comment (alexzherdev) [inline packages/ListBox/src/components/WithTags/WithTags.js:73]: Any benefit to passing `Pill` in parameters vs exporting it from the package? Technically there's no need to pass it in since it's the same every time

File: packages/ListBox/src/components/WithTags/useWithTags.js [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+    const option = func(label);
+    setData(prev => {
+      return prev.concat(option);
+    });
+
+    setSelectedKeys(prev => {
+      return prev.concat(option[key]);
+    });
+
+    setData(prevData => {
```
Comment (alexzherdev) [inline packages/ListBox/src/components/WithTags/useWithTags.js:85]: Why is this necessary in addition to line 77?

File: packages/ListBox/stories/WithTags/examples/CustomTrigger.js [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+import animals from "../mocks";
+
+const defaultFilteredData = animals.slice(0, 20);
+const defaultData = animals;
+
+// prettier-ignore
+const styleForPill = (color) => ({ alignItems: "center", backgroundColor: color.backgroundColor, borderRadius: "50%", boxSizing: "border-box", color: color.fontColor, display: "flex", height: "24px", justifyContent: "center", marginRight: "8px", padding: "3px", width: "24px", fontSize: ".8rem", lineHeight: 1})
+
+export default function CustomTrigger() {
+  const { isSelected, filteredData, getSelectedOptions, ...moreUseListBoxWithTagsProps } = useWithTags("label", {
```
Comment (alexzherdev) [inline packages/ListBox/stories/WithTags/examples/CustomTrigger.js:14]: `useListBoxWithTags`/`Pills` seems a better name, otherwise no idea what exactly I'm using with tags?

File: packages/ListBox/stories/WithTags/examples/Controlled.js [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+      <ListBox
+        noResultsMessage="No results found, but you can add an email and then press enter..."
+        onChange={handleChange}
+        onCustomOption={handleAddedOption}
+        onRemove={handleRemove}
+        selectedOptions={selectedKeys.length ? animals.filter(item => selectedKeys.includes(item.label)) : []}
+        filter={handleFilterDebounce}
+      >
+        {dataFiltered.map(option => {
+          if (typeof option.isCustom !== "undefined") {
```
Comment (alexzherdev) [inline packages/ListBox/stories/WithTags/examples/Controlled.js:93]: This boilerplate is present in all of the stories but it's not clear why it's necessary

File: packages/ListBoxWithTags/src/ListBoxWithTags.styles.js (hunk 9 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/ListBoxWithTags.styles.js]: there is a conversation why having tokens with px and not units creates complexity in some scenarios, I propose to refactor styles so it's possible to do this without "alternatives" / "hacks" / "importing stylers".

For this specific case `tokens.spaceSm` is equal to "4px" which is a string, so I'm just taking the first element which is the 4. https://app.slack.com/client/T02SWPEUM/CB3QRPDJ4/thread/CB3QRPDJ4-1605751958.213700

Comment (nahumzs) [reply packages/ListBoxWithTags/src/ListBoxWithTags.styles.js]: all strings are arrays...

File: packages/ListBoxWithTags/src/components/Pill/Pill.js (hunk 10 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/components/Pill/Pill.js]: yeah I'm fine with this duplication, reads correctly wherever you are and doesn't provide any extra effort that would be adding it to a constant. You know me I'm against micro abstractions.

Comment (nahumzs) [reply packages/ListBoxWithTags/src/components/Pill/Pill.js]: Also the Pill will be their own component so this will get refactor :)

File: packages/ListBoxWithTags/src/helpers.js (hunk 11 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/helpers.js]: yeah in this specific case I used a `for` instead of the `filter` method because theoretically filter it should be slower than a regular `for`, and while 99% of the cases I would be used the `Array.filter` method, in this case, might get call multiple cases in a long list, so didn't add too much effort to me to add this "small" boost of performance even if I haven't been measure it.

File: packages/ListBoxWithTags/src/useListBoxWithTags.js (hunk 13 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/useListBoxWithTags.js]: in case you want to override the way, you create your custom function. 

maybe the following is easier to read?:
```
  const factoryCustomOption = label => ({ label, isCustom: true });
  const handleCustomOption = (func = factoryCustomOption) => label => {
```

File: packages/ListBoxWithTags/test/ListBoxWithTags.spec.js (hunk 16 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/test/ListBoxWithTags.spec.js]: these are not html snapshot these are javascript objects snapshots and are really handy when you want to assert that a specific object is returned whenever running a function. 

Instead of manually written the expected return object just assert that match the snapshot :)

File: packages/ListBox/src/components/WithTags/WithTags.js (hunk 19 above)
Comment (nahumzs) [reply packages/ListBox/src/components/WithTags/WithTags.js:28]: would make any difference? the linter will complain if I use an object isn't it?

File: packages/ListBox/src/helpers/filter.js (hunk 23 above)
Comment (nahumzs) [reply packages/ListBox/src/helpers/filter.js:11]: https://github.com/acl-services/paprika/pull/753#discussion_r529857074

File: packages/ListBox/src/components/WithTags/WithTags.js (hunk 19 above)
Comment (jamiek-acl) [reply packages/ListBox/src/components/WithTags/WithTags.js:28]: I think both behave the same, just one is more concise

File: packages/ListBox/stories/WithTags/examples/Controlled.js (hunk 29 above)
Comment (nahumzs) [reply packages/ListBox/stories/WithTags/examples/Controlled.js:93]: yes, this is for the case when the user is allowed to add custom options via the filter input. 

The `onAddCustomOption()` has a default factory function that look like:

```
(label) => {
  return { label, isCustom: true};
}
```

But can be override as:

```js
onAddCustomOption(label => {
 return { id: nanoid, label, ...yourStuff }
})
```

And delegate the control about how to create this option to the consumer.

I'm planning to add all the full documentation in the next PR with proper `md` and `mdx` documentation.

File: packages/ListBox/src/components/WithTags/useWithTags.js (hunk 27 above)
Comment (nahumzs) [reply packages/ListBox/src/components/WithTags/useWithTags.js:85]:
```js
    const option = func(label);
    
    setData(prev => {
      return prev.concat(option);
    });

    setSelectedKeys(prev => {
      return prev.concat(option[key]);
    });

    setData(prevData => {
      return prevData.concat(option);
    });
```

Seem I applying two times the same line, brain fart.


But this is the full explanation:
```js
// this create the custom options with the filter's input 
    const option = func(label);
// this add the option to the entire data set and force useMemo function to recreate the dictionary line 26
    setData(prev => {
      return prev.concat(option);
    });
// this add the "key" to the selected item
    setSelectedKeys(prev => {
      return prev.concat(option[key]);
    });

// this restart the options filtered to the original defaultFilterData, the initial state.
    setFilteredData(defaultData);
```

File: packages/ListBox/src/components/WithTags/WithTags.js (hunk 26 above)
Comment (nahumzs) [reply packages/ListBox/src/components/WithTags/WithTags.js:73]: Mmm, for this pass it's easier to expose it like this, I will extract the Pill to its own package soon, and then can be imported, don't want to export it as part of the ListBox package.

Meanwhile, this allowed us to expose it easily and also can be fixed quickly whenever we do a breaking change or we can even just import the `Pill` inside of ListBox and still be backward compatible.

File: packages/ListBoxWithTags/src/useListBoxWithTags.js (hunk 15 above)
Comment (nahumzs) [reply packages/ListBoxWithTags/src/useListBoxWithTags.js]: I thought it for the past two hours and agree makes sense, will be more descriptive having the word key on the object.

File: packages/ListBox/stories/WithTags/examples/Controlled.js (hunk 29 above)
Comment (nahumzs) [reply packages/ListBox/stories/WithTags/examples/Controlled.js:93]: I have removed it from the examples because it no longer needs it in most cases.

File: packages/ListBox/src/components/WithTags/WithTags.js (hunk 21 above)
Comment (nahumzs) [reply packages/ListBox/src/components/WithTags/WithTags.js:48]: ohh miss this one, fixing it.

File: packages/ListBox/src/index.d.ts [hunk 30]
```diff
@@ -170,6 +170,28 @@ declare namespace ListBox {
...
+  onChange?: (...args: any[]) => any;
+  /** Callback whenever the user input a new custom option like some@email.com, pass undefined to ignore this behaviour */
+  onAddCustomOption?: (...args: any[]) => any;
+  /** Callback once a pill is remove from the Trigger */
+  onRemove?: (...args: any[]) => any;
+  /** Regex that match the input of the user and reports to onAddCustomOption. The default is a basic email regex */
+  customOptionRegex?: instanceOf;
+  /** Render prop to override the default Pill style, see example for it's uses. */
+  renderPill?: (...args: any[]) => any;
+  /** An array of id that helps the ListBoxWithTags to known what elements are selected */
```
Comment (jamiek-acl) [inline packages/ListBox/src/index.d.ts:192]: A few outdated references to `ListBoxWithTags` in this block

File: packages/ListBox/src/components/Pill/Pill.styles.js [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+    padding: 0;
+  `;
+});
+
+export const Pill = styled.li(() => {
+  return css`
+    align-items: center;
+    background: ${tokens.backgroundColor.level0};
+    border: 1px solid ${tokens.border.color};
+    border-radius: ${tokens.space[0] * 2}px;
```
Comment (jamiek-acl) [inline packages/ListBox/src/components/Pill/Pill.styles.js:21]: These `space[0]` will work... until the token changes to be `10px`

File: packages/ListBox/src/components/Pill/Pill.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+  onRemove: PropTypes.func.isRequired,
+  children: PropTypes.node.isRequired,
+  size: PropTypes.oneOf(["small", "medium", "large"]),
+};
+
+const defaultProps = {
+  size: "medium",
+};
+
+export function Delete({ size, onRemove = () => {} }) {
```
[the remark this answers is not in the document]
Comment (nahumzs) [reply packages/ListBox/src/components/Pill/Pill.js:16]: will be replaced in their own package I think is ok for now.

**Review by jamiek-acl — APPROVED** [review]
tenor
